### PR TITLE
#5: Implemented fallback switch and improved unit tests.

### LINKED ISSUE #5: Make fallback configurable
## Situation

At the moment logging falls back to console logging in case a log connection could not be established. This might not be the right choice for all users and use cases. Instead users should be able to configure that behavior.

## Acceptance Criteria

* Users can set fallback to console
* Users can fallback to no logging
* Users can disable fallback, causing an error

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #5.

### CODE REVIEW

File: .github/workflows/ci-build.yml [hunk 1]
```diff
@@ -15,15 +15,19 @@ jobs:
           sudo apt-get -y install lua${{ matrix.lua_version }}
           sudo apt-get -y install liblua${{ matrix.lua_version }}
           sudo apt-get -y install luarocks
+          sudo luarocks install luacheck
+          sudo luarocks install luacov
+          sudo luarocks install luacov-coveralls
           sudo luarocks install luaunit
           sudo luarocks install mockagne
+          sudo luarocks install luasocket
```
Comment (kaklakariada) [inline .github/workflows/ci-build.yml:23]: Can you combine the separate calls into a single call? Same for apt-get install.

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -129,6 +129,30 @@ else
...
+
+One way you can pick the desired fallback method is via the optional third parameter of the `init` function:
+
+```lua
+log.init(date_pattern, false, log.fallback_strategies.ERROR)
+```
+
+If you omit the third parameter, the default strategy is to fall back to console logging.
+
+Another way is or via `set_fallback_strategy`:
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:150]:
```suggestion
Another way is via `set_fallback_strategy`:
```

File: src/remotelog.lua [hunk 3]
```diff
@@ -39,15 +42,18 @@ local socket = require("socket")
...
 --
 -- @return module loader
 --
-function M.init(timestamp_pattern, use_high_resolution_time)
+function M.init(timestamp_pattern, use_high_resolution_time, fallback_strategy)
     if timestamp_pattern then
         M.timestamp_pattern = timestamp_pattern
     end
     M.use_high_resolution_time = use_high_resolution_time or false
+    M.fallback_strategy = fallback_strategy or fallback_strategies.CONSOLE
```
Comment (kaklakariada) [inline src/remotelog.lua:56]: Duplicate code. Is it possible to call `set_fallback_strategy()` here?

File: src/remotelog.lua [hunk 4]
```diff
@@ -62,6 +68,15 @@ function M.set_client_name(log_client_name)
     M.log_client_name = log_client_name
 end
 
+---
+-- Set the fallback strategy.
+--
+-- @param fallback_strategy what to do if the remote listener connection cannot be established (default: console log)
+--
+function M.set_fallback_strategy(strategy)
```
Comment (kaklakariada) [inline src/remotelog.lua:76]: Is it necessary to have a setter? You can already configure this via `init`. Is it necessary to change the fallback strategy at runtime?
Keep API minimal if possible.

File: tools/runtests.sh [hunk 5]
```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
+
```
Comment (kaklakariada) [inline tools/runtests.sh:2]:
```suggestion
set -o errexit
set -o nounset
set -o pipefail

```

File: tools/runtests.sh [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    mkdir -p "$luacov_dir"
+}
+
+##
+# Run the unit tests and collect code coverage.
+#
+# Return error status in case there were failures.
+#
+function run_tests {
+    cd "$test_module_path" || exit
```
Comment (kaklakariada) [inline tools/runtests.sh:32]: In my opinion a failing `cd` is an error condition. Print a message and `exit 1`

File: tools/runtests.sh [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+readonly reports_dir="$target_dir/luaunit-reports"
+readonly luacov_dir="$target_dir/luacov-reports"
+
+function create_target_directories {
+    mkdir -p "$reports_dir"
+    mkdir -p "$luacov_dir"
+}
+
+##
+# Run the unit tests and collect code coverage.
```
Comment (kaklakariada) [inline tools/runtests.sh:27]:
```suggestion
# Run the unit tests.
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/remotelog.lua (hunk 3 above)
Comment (redcatbear) [reply src/remotelog.lua:56]: Done.

File: src/remotelog.lua (hunk 4 above)
Comment (redcatbear) [reply src/remotelog.lua:76]: Removed setter.

File: tools/runtests.sh (hunk 6 above)
Comment (redcatbear) [reply tools/runtests.sh]: Added non-zero exit code. `cd` already prints a error.

File: .github/workflows/ci-build.yml (hunk 1 above)
Comment (redcatbear) [reply .github/workflows/ci-build.yml]: Done via `for` loop. The tool itself does unfortunately not have this option.

File: doc/user_guide/user_guide.md [hunk 8]
```diff
@@ -129,6 +129,30 @@ else
...
+```
+
+If you omit the third parameter, the default strategy is to fall back to console logging.
+
+Another way is via `set_fallback_strategy`:
+
+```lua
+log.set_fallback_strategies(log.fallback_strategies.ERROR)
+```
+
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:155]:
```suggestion
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**